gsr_booking/booking_logic.py

```diff
@@ -0,0 +1,255 @@
+        bookings = succesful_bookings
+        complete_success = (next_end - next_start < datetime.timedelta(hours=0.1)) and (
+            error is None
+        )
+        partial_success = len(bookings) > 0
+        result_json = {
+            "complete_success": complete_success,
+            "partial_success": partial_success,
+            "rooms": [],
+        }
+
+        # add failed bookings to bookings
+        failed_bookings = self.split_booking(next_start, end, None, False)
+        failed_key = f"{lid}_{room}"
+        if not failed_key in bookings:
+            bookings[failed_key] = []
+        failed_split_bookings = self.split_booking(next_start, next_end, None, False)
+        bookings[failed_key].extend(failed_split_bookings)
+
+        # add successful bookings to result_json
+        for (key, bookings_array) in bookings.items():
+            key_split = key.split("_")
+            lid = key_split[0]
+            room = key_split[1]
+            result_json["rooms"].append({"lid": lid, "room": room, "bookings": bookings_array})
+
+        # handle potential error
+        if error is not None:
+            if "not a valid".lower() in error.lower():
+                result_json["error"] = "Attempted to book for an invalid time slot"
+            elif "daily limit".lower() in error.lower():
+                result_json["error"] = "Group does not have sufficient booking credits"
+            else:
+                result_json["error"] = str(error)
+        return result_json
+
+    def get_used_booking_credit_for_user(self, lid, email):
+        # returns a user's used booking credit (in hours) for a specific building (lid)
+        RESERVATIONS_URL = "https://api.pennlabs.org/studyspaces/reservations"
+        if lid == "1":
+            return (False, 0)  # doesn't support huntsman yet
+        try:
+            r = requests.get(RESERVATIONS_URL + "?email=" + email)
+            if r.status_code == 200:
+                resp_data = r.json()
+                reservations = resp_data["reservations"]
+                used_credit_hours = 0
+                for reservation in reservations:
+                    from_date = datetime.datetime.strptime(
+                        reservation["fromDate"], self.DATE_FORMAT_STR
+                    )
+                    to_date = datetime.datetime.strptime(
+                        reservation["toDate"], self.DATE_FORMAT_STR
+                    )
+                    reservation_hours = (to_date - from_date).total_seconds() / 3600
+                    if str(reservation["lid"]) == lid and reservation_hours > 0.1:
+                        used_credit_hours += reservation_hours
+                return (True, used_credit_hours)
+        except requests.exceptions.RequestException as e:
+            print(e)
+        return (False, 0)
+
+    def split_booking(self, start, end, pennkey, booked):
+        # splits a booking into smaller bookings (of min_slot_hrs) for displaying to user
```

Comment:
```suggestion
        """
        splits a booking into smaller bookings (of min_slot_hrs) for displaying to user
        """
```
Docstrings in python are triple-quoted strings :)